### DISCUSSION TITLE
nested replies + creating constraint for allowing only 1 of 2 columns to have value

### DISCUSSION THREAD

**Discussion by TamirCode** [prisma/prisma#19701] [Q&A] [answered]
I will have users, posts, and comments. 
The comments will be recursive (infinite nested replies).
![image](https://github.com/prisma/prisma/assets/66864667/3d0f6278-3c90-4659-8a4e-be4d1974bac7)

Here is my schema:
```
model User {
    id        Int      @id @default(autoincrement())
    username  String   @unique
    password  String
    settings  Json?
    role      Role     @default(USER)
    createdAt DateTime @default(now())
    updatedAt DateTime @updatedAt

    // one to many
    posts    Post[]    @relation("posts")
    comments Comment[] @relation("comments")

    // many to many
    likedPosts       Post[]    @relation("likedPosts")
    dislikedPosts    Post[]    @relation("dislikedPosts")
    likedComments    Comment[] @relation("likedComments")
    dislikedComments Comment[] @relation("dislikedComments")
}

model Post {
    id        Int      @id @default(autoincrement())
    content   String
    createdAt DateTime @default(now())
    updatedAt DateTime @updatedAt

    // many to one
    user     User      @relation("posts", fields: [userId], references: [id])
    userId   Int
    comments Comment[]

    // many to many
    likes    User[] @relation("likedPosts")
    dislikes User[] @relation("dislikedPosts")
}

model Comment {
    id        Int      @id @default(autoincrement())
    content   String
    createdAt DateTime @default(now())
    updatedAt DateTime @updatedAt

    // one to many
    user   User  @relation("comments", fields: [userId], references: [id])
    userId Int
    post   Post? @relation(fields: [postId], references: [id])
    postId Int?

    // many to many
    likes    User[] @relation("likedComments")
    dislikes User[] @relation("dislikedComments")

    // one to many self-reference
    childComments   Comment[] @relation("replies")
    // many to one self-reference
    parentComment   Comment?  @relation("replies", fields: [parentCommentId], references: [id])
    parentCommentId Int?
}

enum Role {
    USER
    ADMIN
}
```

Inside comment model, I want my prisma to disallow having both a `parentCommentId` and `postId` because there can only be one or the other.

I am open to hearing any other tips/suggestions and different ways to create nested replies, thanks!

**Comment by nurul3101** [comment, marked as answer]
Hello @TamirCode 👋 

There are two ways through which you can enforce this validation:

1. **At Application level**:  When you create or update a comment in your application, you can add a check to ensure that only one of `parentCommentId` or `postId` is set.

Here's an example code snippet:

```typescript
async function createComment(data: Partial<Comment>) {
  if (data.parentCommentId != null && data.postId != null) {
    throw new Error('A comment cannot have both a parentCommentId and a postId');
  }
  if (data.parentCommentId == null && data.postId == null) {
    throw new Error('A comment must have either a parentCommentId or a postId');
  }
  return prisma.comment.create({
    data,
  });
}
```

You can also use [Prisma Client Extension](https://www.prisma.io/docs/concepts/components/prisma-client/client-extensions) to do these kinds of validations, here is an [example](https://github.com/prisma/prisma-client-extensions/tree/main/input-validation).

2. **At database level**:  You can define these constraints directly in the database. For example, if you're using PostgreSQL, you might add a check constraint like this:

```sql
ALTER TABLE Comment
ADD CONSTRAINT check_parent_or_post
CHECK (
  (parent_comment_id IS NOT NULL)::integer +
  (post_id IS NOT NULL)::integer = 1
);
```

This constraint ensures that exactly one of parent_comment_id and post_id is not null. 

Hope this helps! Please feel free to let us know in case you have any additional queries. If not, please mark this discussion as Answered so that we are aware that this is resolved.

**Reply by nurul3101** [reply]
I am checking to see if you had a chance to look at my last response.

It would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 
Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by nurul3101** [reply]
@TamirCode Do let us know in case you still need any assistance.
